### PR TITLE
CI and release improvements

### PR DESCRIPTION
**Pull request by codeadict**
## Proposed Changes

**CI Improvements:**

* Test on the last 3 major OTP versions, following the Erlang/OTP support conventions.
* Test on Both Linux and OSX for each version
* Report test failures with JUnit, allowing to identify flaky tests
* Smoke test releases on all OSes and Erlang/OTP versions and add more extensive smoke test by publishing/subscribing to messages.

**Release Improvements:**

* Fixes release configuration problems in OSX in OTP 23 and 24. Fixes https://github.com/vernemq/vernemq/issues/1926
* Fixes release configuration problems when using Unix domain sockets. Fixes https://github.com/vernemq/vernemq/issues/2008, https://github.com/vernemq/vernemq/issues/1634
* Upgrade to Rebar3 1.19.x
* Releases can now build on OTP 25
* Releases can now build on ARM architectures (Tested on M1 Mac and Raspberry PI) Fixes https://github.com/vernemq/vernemq/issues/1421 and partially https://github.com/vernemq/docker-vernemq/issues/173 and https://github.com/vernemq/docker-vernemq/issues/340 (Requires multi-arch image)

**Maintenance Improvements**

* Use new OTP 25 peer module to allow tests running in OTP25.
* Removes any master/slave terminologies for more inclusive language. Still the main branch is `master`, we probably should switch to `main`.
* Fixes some flaky tests, skipped some in CI, yet there are some flaky tests that are repoprted in GHA, need to aim for 0.
* Fix some unused variable warnings

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [X] Bugfix (non-breaking change which fixes issue #XXXX)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, styles...)
- [X] DevOps (Build scripts, pipelines...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating
the PR. If you're unsure about any of them, don't hesitate to ask on the
mailing list. We're here to help! This is simply a reminder of what we are
going to look for before merging your code._

- [X] I have read the `CODE_OF_CONDUCT.md` document
- [X] All tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if needed)
- [ ] Any dependent changes have been merged and published in related repositories
- [ ] I have updated changelog (At the bottom of the release version)
- [X] I have squashed all my commits into one before merging

## Further Comments

@ioolkos - Opening this for early reviews, still need the main cuttlefish repo to accept my PR https://github.com/Kyorai/cuttlefish/pull/33 and update the changelog plus this PR description.

### CODE REVIEW

**Comment by ioolkos** [comment]
@codeadict :tada: :tada: :1st_place_medal: :tada: :tada:

**Comment by codeadict** [comment]
@ioolkos You need to retry the failed jobs as I don't have permissions to do that. Another approach is i probably should retry failed tests til they pass

**Comment by ioolkos** [comment]
@codeadict anything more to adapt, now that your Cuttlefish PR is merged? 
Btw, yes, let's of course switch to `main` branch naming.

**Comment by codeadict** [comment]
@ioolkos - I will point to the official cuttlefish now. Need to also make some changes to the Junit reporter, it works well on my fork but when the PR to the main repo comes from a fork, it gets messed up. Also want to somehow retry the CT failed tests and then we can merge see if we can merge this

**Comment by codeadict** [comment]
@ioolkos I don't know how the packaging scripts will work if I rename the branch to main since this is not Open Source. You will need to look into that and make sure renaming master to main will not break things

**Comment by ioolkos** [comment]
@codeadict okay, let me know when I should look into any specific test.
Renaming is fine, I'll adapt the package builders.

**Comment by codeadict** [comment]
@ioolkos we would need to merge https://github.com/vernemq/rebar3_cuttlefish/pull/45 too

**Comment by ioolkos** [comment]
Build currently fails for me on Ubuntu/OTP 25.2:

```
===> Linking priv/eleveldb.so
===> sh(cc c_src/eleveldb.o c_src/refobjects.o c_src/workitems.o $LDFLAGS $LDLIBS -shared  -L"/home/il/Erlangs/25.2/lib/erl_interface-5.3/lib" -lei c_src/leveldb/libleveldb.a c_src/system/lib/libsnappy.a -lstdc++ -o priv/eleveldb.so)
failed with return code 1 and the following output:
/usr/bin/ld: c_src/system/lib/libsnappy.a(snappy.cc.o): relocation R_X86_64_PC32 against symbol `_ZTVN6snappy4SinkE' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status

make: *** [Makefile:22: rel] Error 1
```

**Comment by ioolkos** [comment]
@codeadict ah, thanks, had to remove libsnappy-dev. I went down a totally wrong rabbit hole, verifying the library builds etc. They are correctly built with `-fPIC`.

**Comment by ioolkos** [comment]
That libsnappy thing still is a little unfortunate. While it works to remove libsnappy-dev on one of my Ubuntu system (where I had no previous Verne package install), on the other it doesn't. I could try to remove `libsnappy1v5` fully. But this will remove a ton of packages, with VerneMQ among those :)

I need to investigate this. It should somehow be possible to link to a snappy lib built into the `_build` dir somewhere.

**Comment by codeadict** [comment]
Indeed @ioolkos , there must be a way to ignore the OS snappy completely no no rely on it to compile VerneMQ

**Comment by codeadict** [comment]
@ioolkos I've opened https://github.com/basho/eleveldb/issues/270 for the Snappy issue. I'll do more debugging on that and see how we can improve. Looks like the PR there solves the problem in Linux but still breaks for me on OSX

**Comment by ioolkos** [comment]
@codeadict fastastic, thanks!

**Comment by codeadict** [comment]
Hopefully we can get https://github.com/basho/eleveldb/pull/272 merged and this should be ready 🤞🏽

**Comment by codeadict** [comment]
@ioolkos - Should we fork the basho repo from the newest code and add my branch there? We should also get rid of eLevelDb at some point soon. Both the Basho and the Google LevelDb repos have notices to be essentially in maintenance mode. Basho barely has contributors that will make any significant fixes. Seems like after this we probably should focus effort on Rocks DB since it's very well maintained by Facebook and the same on the Erlang library, Benoit and other contributors are really on top of it.

**Comment by ioolkos** [comment]
@codeadict Yes, RocksDB was always the plan to go forward :)
Side note: So far, I see a suprisingly low number of issues coming from LevelDB/eleveldb. But we absolutely need RocksDB too.
We could have a build target that builds all of LevelDB, RocksDB and Leveled so that users can choose, and/or specific 
DB builds.

We can fork eleveldb again, of course, should Basho not merge your Snappy build fixes.

**Comment by ioolkos** [comment]
@codeadict if you have a moment, could you please rebar-lock to the latest basho eleveldb commit? did you plan on upgrading other deps?

**Comment by codeadict** [comment]
Ill do that today @ioolkos . Was sick yesterday and didnt get close to the computer much

**Comment by ioolkos** [comment]
> Was sick yesterday and didnt get close to the computer much

Oi, get well! :pill: 
no hurry here.

**Comment by ioolkos** [comment]
@codeadict haven't yet looked into that test; I think it works in main branch (where it's more the `cluster_leave` one that fails).

**Comment by ioolkos** [comment]
Just did a test run with the newly released OTP 25.3. 
All test passing, except some totally new stuff failing in `vmq_schema_SUITE`. I'll check what's repeteable. Fun :)

**Comment by ioolkos** [comment]
Somehow the Cuttlefish config generation is now broken.

```
➜  vernemq git:(2078) _build/default/rel/vernemq/bin/vernemq config generate -l
escript: exception error: bad argument
  in function  list_to_atom/1
     called as list_to_atom(undefined)
     *** argument 1: not a list
  in call from cuttlefish_escript:main/1 (/home/afa/Erlang/testing/2078/vernemq/_build/default/lib/cuttlefish/src/cuttlefish_escript.erl, line 80)
  in call from escript:run/2 (escript.erl, line 750)
  in call from escript:start/1 (escript.erl, line 277)
  in call from init:start_em/1 
  in call from init:do_boot/3 
```

**Comment by ioolkos** [comment]
Deleted previous comment, what I meant:

```
➜  vernemq git:(2078) _build/default/rel/vernemq/bin/vernemq config generate -l debug
...
2023-03-08T21:22:44.723738+01:00 [error] Error generating configuration in phase apply_translations
2023-03-08T21:22:44.723787+01:00 [error] Error running translation for vmq_server.listeners, [error, {badmatch, {[{{ok, {127,0,0,1}}, 8888}], [{listener_name, "default"}]}}] 
```

File: apps/vmq_server/src/vmq_schema.erl [hunk 1]
```diff
@@ -532,7 +532,7 @@ extract(Prefix, Suffix, Val, Conf) ->
     NameSubPrefix = lists:flatten([Prefix, ".$name"]),
     [
         begin
-            Addr = parse_addr(StrAddr),
+            {ok, Addr} = parse_addr(StrAddr),
```
Comment (ioolkos) [inline apps/vmq_server/src/vmq_schema.erl:535]: We need to add `{ok, Addr} = parse_addr(StrAddr)` to the `extract_var` function (added by @mths1), a couple of lines above.
This fixes the schema generation issue.

**Comment by ioolkos** [comment]
Locally, passing all tests on OTP 25. Nice.

**Comment by ioolkos** [comment]
@codeadict reran the macos/otp25 test in CI, passing. Do you know of a way to take out apps from the coverage test? (see test results from coverage)

**Comment by ioolkos** [comment]
@codeadict thanks, let's see after the current test run is through. I think previously the cover job errored about not finding a test suite for vmq_passwd.

**Comment by codeadict** [comment]
Yeah OTP 25 on mac failed, tests need to be stabilized a bit but it's good to have CI back even if not perfect

**Comment by ioolkos** [comment]
@codeadict very cool. Hope the macos thing is not an OS architecture issue (works with OTP < 25, though). Maybe we can also test a Linux version for a ARM64 architecture, if that's possible.

**Comment by codeadict** [comment]
@ioolkos - Definitely not an OS architecture problem, it's more than networking on OSX is slower and some tests fail but they pass when retried. We can definitely test on Raspbian  ARM64 using Docker but I would prefer to do that on a separate PR. After this PR is merged, we will need to figure out what we need to do to add multi-arch docker images so we can run verneMQ on M1 macs.  I know the packaging/release scripts are not public but would be great to have a CI workflow for releases that create the packages/binaries (Including prebuilt for OSX so we can update the Homebrew package), then when the release workflow finishes it "notifies" docker-vernemq to release the containers and Helm charts automatically.

All this is doable without exposing the packaging scripts if they are in a private repo, we can give GitHub actions a token that has read access on the scripts repo. We can talk about it via chat.

**Comment by codeadict** [comment]
@ioolkos - Id suggest canceling the pending check and re-running it. Sometimes Github VMs get stuck and don't go anywhere, which happens often even in non-Erlang projects I've seen it.

**Comment by ioolkos** [comment]
@codeadict I'm going to merge, in case you don't have any last minute objection :)

**Pull request merged**